Handle encrypted files in "import key" file selection

runnerway opened the pull request:
This is a solution for:
https://github.com/open-keychain/open-keychain/issues/1673

runnerway commented:
The only thing left is that some activity need to close themselves after the operation is completed.
I will address this probably tomorrow. Then it will be ready.

runnerway commented:
Now it works pretty well.
The user experience is good and all seems in place.

Valodim commented:
Awesome lookin' PRs. They also aren't trivial to review, please excuse the delay. I'll try to get to them tomorrow :+1:

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java:
```diff
@@ -104,20 +124,133 @@ public void onAttach(Activity activity) {
+            try {
+                br.close();
+            } catch (IOException e) {
+                e.printStackTrace();
```
Log.e

runnerway replied:
Yes. Will fix it.

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java:
```diff
@@ -104,20 +124,133 @@ public void onAttach(Activity activity) {
+                isEncrypted = new String(buffer).equals(header);
+            }
+        } catch (IOException e) {
+            throw e;
```
isn't this a nop? missing a Log.e mayhaps?

runnerway replied:
Yes. Will fix it.

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java:
```diff
@@ -104,20 +124,133 @@ public void onAttach(Activity activity) {
+
+        if (isEncrypted) {
+            Intent intent = new Intent(getActivity(), DecryptActivity.class);
+            intent.setAction(DecryptActivity.ACTION_DECRYPT_FILE_AND_RETURN);
```
I think this should be handled directly as part of the decryption operation, no need to roundtrip. That way we get the same behavior with encrypted keys regardless of what activity the input comes from and the activity stack. Just `startActivity`, `finish` and proceed from there.

runnerway replied:
Yes. That was my first try, but i spent a lot of time on it with not very good results regarding to user experience.
I thought the roundtrip was better for some reasons.

But i've just had some new ideas, i will try them and report back.
Thank you for the review.

runnerway commented:
I've fixed the little problems you pointed out and i tried again to do without roundtrip.
Updated PR to new version. (I still have the older PR on my pc if we want to go back to that)

We have mainly 2 problems:
1) We can't "finish" the activity because otherwise we lose the permission to access the Uri
(http://developer.android.com/guide/topics/providers/content-provider-basics.html#Intents)
2) We need to close all the activities opened during the process to import encrypted keys, as soon as the operation is finished.

So i propose to do as in the latest version of this PR.
After we finish importing new keys, go back to MainActivity. (If not specified otherwise)
This would solve problem n.2 and it makes sense to me. 
Probably the user will go back to MainActivity to see the just imported keys anyway. 

I'm a bit tired now. I hope i didn't forget anything important.

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java:
```diff
@@ -419,6 +419,16 @@ protected void handleResult(ImportKeyResult result) {
         }
 
+        if (result.isOkNew() || result.isOkUpdated()) {
+            Preferences prefs = Preferences.getPreferences(this);
```
extract into method please

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java:
```diff
@@ -419,6 +419,16 @@ protected void handleResult(ImportKeyResult result) {
+
+            Intent intent = new Intent(this, MainActivity.class);
+            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
+            startActivity(intent);
```
should this `finish()`?

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java:
```diff
@@ -121,4 +139,102 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
+        }
+    }
+
+    private boolean isEncryptedFile(Uri uri) throws IOException {
```
make static and move to `FileHelper` please

Valodim commented:
Good point about the uri permissions. Just tested it, look and feel seems good to me :+1:

runnerway commented:
I will submit an updated PR with your notes fixed later today.
Thank you for the review.

runnerway commented:
This should be ready for review. :)

dschuermann commented:
Looks good. @Valodim : merge?

Valodim commented:
Yup. I tested this before and code looks good now.
